### ISSUE TITLE
Yolo v8 what does deterministic parameter do?

### ISSUE THREAD

**Issue by Diogo-Valente2111** [ultralytics/ultralytics#11210]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Yolo v8 what does deterministic parameter do? The definition is: Forces deterministic algorithm use, ensuring reproducibility but may affect performance and speed due to the restriction on non-deterministic algorithms.
I'm wondering what are the non-deterministic algorithms that are being suppressed.

**Comment by glenn-jocher** [comment]
Hello! The `deterministic` parameter in YOLOv8 ensures that when enabled, the model will consistently produce the same output for the same input, which is crucial for reproducibility in experiments. Non-deterministic behaviors typically arise from certain GPU operations and multi-threading data loading techniques that can yield minor variations in results during different runs, even with the same inputs. In deep learning, common sources of non-determinism include certain convolutional operations and specific types of data augmentation. Opting for deterministic behavior might inhibit the model's ability to leverage these faster, non-deterministic operations, potentially leading to slower processing times. Hope this clarifies your doubt! 😊

**Comment by Petros626** [comment]
@glenn-jocher is it recommended to use `deterministic=True` for training? It affects the performance during training, inference or both?

**Comment by glenn-jocher** [comment]
@Petros626 using `deterministic=True` ensures reproducibility by enforcing deterministic algorithms but may slightly reduce performance during training due to restricted optimizations. It does not affect inference performance. Use it if reproducibility is critical for your experiments. For more details, refer to the deterministic settings in the [Ultralytics documentation](https://docs.ultralytics.com/usage/cfg/).

**Comment by rauffatali** [comment]
If I want to run yolo model multiple times with different seeds (to see how results varies across different runs)  for benchmarking purposes, should I turn off this argument or keep it (as True)? Thanks.

**Comment by glenn-jocher** [comment]
Keep deterministic=True and sweep only the `seed` so any variation is attributable to the seed (for stricter control, set `workers=0`); disable it only if you intentionally want to include kernel-level nondeterminism in your variance. For convenience:
```bash
for s in 0 1 2 3 4; do
  yolo train model=yolo11n.pt data=coco8.yaml epochs=100 seed=$s deterministic=True workers=0
done
```
See the `deterministic` and `workers` args in the Train settings and how seeding is applied in the init_seeds reference.

**Comment by Petros626** [comment]
@glenn-jocher I got a question. 

When I got one default dataset and further 3 datasets (same data, but used with one data augmentation technique (densification, interpolation, upsampling)) is it then recommended to set `deterministic=True` to get same results or even better?

**Comment by glenn-jocher** [comment]
`deterministic=True` won’t make results “better”; it just makes each run repeatable so you can attribute differences to your dataset/augmentation changes rather than RNG noise—so for a fair comparison across your 4 dataset variants, keep `deterministic=True`, fix `seed` (and for strict repeatability set `workers=0`), then compare results across several seeds per dataset (mean/std) instead of relying on a single run; the seeding behavior is implemented in [`init_seeds()`](https://docs.ultralytics.com/reference/utils/torch_utils/) and the dataloader worker seeding in [`seed_worker()`](https://docs.ultralytics.com/reference/data/build/), and since this issue is closed/stale please drop any follow-ups in [Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions) for visibility.

**Comment by basilevs** [comment]
Beware that even deterministic detection result depends on batch contents:

```
"""
Demonstrates detection non-determinism depending on input length
"""

from pathlib import Path
from ultralytics import YOLO
from ultralytics.engine.results import Results
from ultralytics.utils.torch_utils import init_seeds

model = YOLO('runs/detect/train35/weights/last.pt')

root = Path('data/images')
file_of_interest = root / 'Val' / '2025-10-10_001027.png'
side_effect_file = root / 'Train' / '2026-01-27_000003.jpeg'

def is_expected(files: list[Path]):
    init_seeds(0, True)
    assert file_of_interest in files
    resultsIterator = model.predict(files, stream=True, verbose=False, workers=0, deterministic=True)
    result = next(i for i in resultsIterator if Path(i.path) == file_of_interest)
    return 0 in result.boxes.cls

assert is_expected([file_of_interest]) # pass
assert is_expected([file_of_interest, side_effect_file]) # fail
```

In my case, a specific image of strange dimensions corrupts detection for its batch neighbors (but only for some classes)

**Comment by glenn-jocher** [comment]
Thanks for the repro—this kind of “neighbor image” effect usually means the *preprocessing path* changed when you switched from a single image to a mixed-shape batch (e.g., `LetterBox(...)` behavior can depend on batch shape/`rect` inside [`BasePredictor.pre_transform()`](https://docs.ultralytics.com/reference/engine/predictor/)), so the same file can be resized/padded differently and legitimately produce different detections; could you please open a fresh bug report via the [new issue form](https://github.com/ultralytics/ultralytics/issues/new/choose) with your full `yolo checks` output and the original `(H, W)` of both images (or a public repro), and as a quick sanity check try forcing fixed preprocessing like this:

```python
results = model.predict(
    [file_of_interest, side_effect_file],
    stream=True,
    imgsz=640,
    rect=False,
    workers=0,
    verbose=False,
)
```

**Comment by basilevs** [comment]
Rect=False fixes the specific test I used. I'm retesting on my whole dataset.

**Comment by glenn-jocher** [comment]
Good find—if you need per-image results to be invariant to what else is in the batch, keep `rect=False` (and fix preprocessing with a constant `imgsz`; optionally set `batch=1` to remove batch effects entirely); if anything still changes across your full-dataset retest, please open a new issue with a small public repro following the [Minimum Reproducible Example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/) and include the two original images (and their `(H, W)` sizes).
